packages/orama/src/components/algorithms.ts

```diff
@@ -41,16 +49,47 @@ export function prioritizeTokenScores(arrays: TokenScore[][], boost: number, thr
     return results
   }
 
+  // Prepare keywords count tracking for threshold handling
+  const allResults = results.length
+  const tokenKeywordsCount: [InternalDocumentID, number][] = []
+
+  for (const tokenKeywordsCountEntry of tokenKeywordsCountMap.entries()) {
+    tokenKeywordsCount.push(tokenKeywordsCountEntry)
+  }
+
+  // Find the index of the last result with all keywords.
+  // Note that since score is multipled by 1.5 any time the token is encountered in results it means
+  // that tokenMap and tokenKeywords should always have the same order.
+  const keywordsPerToken = tokenKeywordsCount.sort((a, b) => b[1] - a[1])
+
+  let lastTokenWithAllKeywords: number | undefined = undefined
+  for (let i = 0; i < allResults; i++) {
+    if (keywordsPerToken[i][1] === keywordsCount) {
+      lastTokenWithAllKeywords = i
+    } else {
+      break
+    }
+  }
+
+  // If no results had all the keywords, either bail out earlier or normalize
+  if (typeof lastTokenWithAllKeywords === 'undefined') {
+    if (threshold === 0) {
+      return []
+    }
+
+    lastTokenWithAllKeywords = 0
+  }
+
   // If threshold is 0, it means we will only return all the results that contains ALL the search terms (exact match)
   if (threshold === 0) {
-    const shortestArrayLength = Math.min(...arrays.map(arr => arr.length))
-    return results.slice(0, shortestArrayLength)
+    return results.slice(0, lastTokenWithAllKeywords + 1)
   }
 
   // If the threshold is between 0 and 1, we will return all the results that contains at least the threshold of search terms
   // For example, if threshold is 0.5, we will return all the results that contains at least 50% of the search terms
   // (fuzzy match with a minimum threshold)
-  const thresholdLength = Math.ceil((threshold * 100 * results.length) / 100)
+  const thresholdLength =
+    lastTokenWithAllKeywords + Math.ceil((threshold * 100 * (results.length - lastTokenWithAllKeywords)) / 100)
```

Comment:
Are there some tests that test this? In particular, the comment above.

Reply:
Yup, see: https://github.com/oramasearch/orama/blob/improved-sort/packages/orama/tests/threshold.test.ts#L128